### ISSUE TITLE
*bug* (spreadsheet) scroll is too fast on mac touchpad

### ISSUE THREAD

**Issue by RomaricMollard** [ONLYOFFICE/DocumentServer#244]
What is the current behavior?
Scroll is too fast, it's impossible to work with the touchpad scroll (both vertically and horizontally).
I suppose you use the scroll 'tick' (or an "onScroll") to jump to the next row/column, but I think it will work well on mac touchpad reasoning with pixels width and height (of both cells and touchpad scroll deltas)
Romaric

**Comment by ShockwaveNN** [comment]
Could you specify your Mac device, Mac OS version and version of DocumentServer

**Comment by RomaricMollard** [comment]
Yes, sorry :
Issue observed on
- Sierra and High sierra,
- Mac Book pro 2011 and 2017,
- DocumentServer version 5.0.6

**Comment by ShockwaveNN** [comment]
@RomaricMollard Yeah, thank you, I can confirm this is a bug.
Issue 37109 in our internal issue tracked, we try to fix this behavior in future releases.

**Comment by pciavald** [comment]
Hi, using OnlyOffice spreadsheet scroll is quite hard on Linux and Windows too on our side. We think that the best option would be to provide a choice to scroll using pixels OR rows/columns. Thank you !

**Comment by Rita-Bubnova** [comment]
Hello, @pciavald.
Thank you  for the idea. I created a new issue 39100 in our private issue tracker.
We will discuss this with our team.

**Comment by abdyfranco** [comment]
I'm having the same issue on a Mac Pro using a Magic Trackpad 🙁

**Comment by RomaricMollard** [comment]
We bought the complete edition and the bug is here too. It’s almost impossible to use spreadsheet with a touchpad or a Magic Mouse... 😕

**Comment by RomaricMollard** [comment]
Hello, I saw "Enhanced completely rewritten scroll" for spreadsheet in the Changelogs of v5.2.3 !
I updated our installation expecting a new better scroll in spreadsheet, but i'm experiencing exactly the same results as before, what exactly did you change @ShockwaveNN @Rita-Bubnova in the last version ?
I assume is wasn't related to this thread, as it is not closed yet.

**Comment by ShockwaveNN** [comment]
@RomaricMollard yep, you're right, this changes are not related to this exact problem

**Comment by martinspaeth** [comment]
Hey all together, thank you for your awesome work!
We started using onlyoffice as an extension to nextcloud and it looks very promissing. 
We use macs all over the company and facing the same issue. We use ce but plan to switch to a complete edition soon. Facing the same problem using chrome and safari is a blocker so.

I think the problem is related to event fired in the browser. Since the trackpad/magic mouse is very sensitive a scroll never only represents  a scroll up/down but also a scrolling to the left/right. 
Two problems sum up here i guess:
1. Since the mac scrolling is not linear but adds accelleration a small change may be seen with a big representation in the event
2. The scrolling in sheets is implemented cell by cell not pixel by pixel. a small scroll to the right means a scroll to the right by one cell.

Looking at google sheets they fixed this by using a very big threashold on change until a scroll event is triggered.

Hope this helps looking forward for you response

**Comment by martinspaeth** [comment]
I monkey patched our instance by adding a scroll buffer for both scroll directions into the cells events controller.

I inserted the following lines before `this.handlers.trigger("updateWorksheet", /*x*/undefined, /*y*/undefined, /*ctrlKey*/undefined,` at line 1633:

https://github.com/ONLYOFFICE/sdkjs/blob/master/cell/view/EventsController.js#L1633

```
 
if (!this.xScrollBuffer) { this.xScrollBuffer = 0; }
if (!this.yScrollBuffer) { this.yScrollBuffer = 0; }

this.yScrollBuffer += deltaY;
this.xScrollBuffer += deltaX;

if (Math.abs(this.xScrollBuffer) > 1) { 
  // allow event to fire for x axis
  deltaX = this.xScrollBuffer; 
  // reset the scroll buffer
  this.xScrollBuffer = 0; 
} else {
  // buffer not reached the threshold yet.. prevent event to be fired
  deltaX = 0; 
}
if (Math.abs(this.yScrollBuffer) > 1) { 
     // allow event to fire for y axis
   deltaY = this.yScrollBuffer; 
     // reset the scroll buffer
   this.yScrollBuffer = 0; 
} else { 
     // buffer not reached the threshold yet.. prevent event to be fired
   deltaY = 0; 
}

// XXX override wheelScrollLinesV to 1 to be able to scroll line by line
self.settings.wheelScrollLinesV = 1;
```


This buffers scrolling in each direction. A scroll happens only after reaching the threshold of 1.
This results in a very smooth nice scrolling on mac (not tested in other environments so).

**Comment by martinspaeth** [comment]
Small addition: The same scrolling issues also applies to slides

**Comment by martinspaeth** [comment]
We use this patch for over a month seems to work fine. What is the process to get this into the next patch?

**Comment by ShockwaveNN** [comment]
Hi, our team implemented fix at
https://github.com/ONLYOFFICE/sdkjs/commit/5e1191fafebc7fb85e5aaa684cdc2067325587a1
https://github.com/ONLYOFFICE/sdkjs/commit/63e15e58ac3a757aa95c22f4a3400d717f8442f8
and it will be release as 5.3

**Comment by ShockwaveNN** [comment]
DocumentServer 5.3.0 is released. This issue has been fixed.
